modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -0,0 +1,799 @@
+        }
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @param res Results.
+     * @param errs Errors.
+     */
+    private void finishPrepare(UUID reqId, Map<UUID, ArrayList<StoredCacheData>> res, Map<UUID, Exception> errs) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (ctx.clientNode())
+            return;
+
+        Exception failure = checkFailure(errs, opCtx0, res.keySet());
+
+        if (failure == null) {
+            assert opCtx0 != null : ctx.localNodeId();
+
+            Map<Integer, StoredCacheData> globalCfgs = new HashMap<>();
+
+            for (List<StoredCacheData> storedCfgs : res.values()) {
+                if (storedCfgs == null)
+                    continue;
+
+                for (StoredCacheData cacheData : storedCfgs)
+                    globalCfgs.put(CU.cacheId(cacheData.config().getName()), cacheData);
+            }
+
+            opCtx0.cfgs = globalCfgs;
+
+            if (U.isLocalNodeCoordinator(ctx.discovery()))
+                cacheStartProc.start(reqId, reqId);
+
+            return;
+        }
+
+        if (opCtx0 == null)
+            finishProcess(failure);
+        else // Remove files asynchronously.
+            ctx.cache().context().snapshotMgr().snapshotExecutorService().execute(() -> {
+                rollback(opCtx0);
+
+                finishProcess(failure);
+            });
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @return Result future.
+     */
+    private IgniteInternalFuture<Boolean> cacheStart(UUID reqId) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (opCtx0 == null)
+            return new GridFinishedFuture<>();
+
+        if (!reqId.equals(opCtx0.reqId)) {
+            return new GridFinishedFuture<>(
+                new IgniteCheckedException("Unknown snapshot restore operation was rejected."));
+        }
+
+        if (!U.isLocalNodeCoordinator(ctx.discovery()))
+            return new GridFinishedFuture<>();
+
+        DiscoveryDataClusterState state = ctx.state().clusterState();
+
+        if (state.state() != ClusterState.ACTIVE || state.transition())
+            return new GridFinishedFuture<>(new IgniteCheckedException(OP_REJECT_MSG + "The cluster should be active."));
+
+        Throwable err = opCtx0.err.get();
+
+        if (err != null)
+            return new GridFinishedFuture<>(err);
+
+        if (!allNodesInBaselineAndAlive(opCtx0.nodes))
+            return new GridFinishedFuture<>(new IgniteCheckedException(OP_REJECT_MSG + "Server node(s) has left the cluster."));
+
+        GridFutureAdapter<Boolean> retFut = new GridFutureAdapter<>();
+
+        try {
+            Collection<StoredCacheData> ccfgs = opCtx0.cfgs.values();
+
+            // Ensure that shared cache groups has no conflicts before start caches.
+            for (StoredCacheData cfg : ccfgs) {
+                if (!F.isEmpty(cfg.config().getGroupName()))
+                    ensureCacheAbsent(cfg.config().getName());
```

Comment:
I think the message which is thrown here is incorrect since you already have a guarantee that cache doesn't exist here. And if it does it means you violate your process guarantees - probably it's better the `IgniteIllegalStateException` throw here.

Reply:
Changed to `IgniteIllegalStateException`